Improve BLTOUCH_HEATERS_OFF functionality

bgort opened the pull request:
- rename to SILENT_PROBING and add SILENT_HEATERS, SILENT_FANS to increase flexibility
 - move heater pausing functionality into thermalManager
   - add variables, pause(), ispaused(), other functions
 - add fan pausing functionality
 - move pausing into set_probe_deployed() so other probe types can benefit
 - handle paused state in M104, M106, M107, M109, M140, and M190

Ref #6606 (this deals with the kill-related issue presented there and implements some of what's discussed).

This has been preliminarily tested and seems to work as intended.  

I believe it's 'safe', but more testing and feedback would be good!

bgort committed:
Improve BLTOUCH_HEATERS_OFF functionality
 - rename to PROBING_HEATERS_OFF
 - move heater pausing functionality into thermalManager
   - add variables, pause(), ispaused(), other functions
 - add fan pausing functionality -> PROBING_FANS_OFF
 - add probing_pause() wrapper
 - move pausing into do_homing_move() and do_probe_move() to minimize quiet time and so other probe types can benefit
 - example configs

bgort commented:
Also, may be able to unify some of the features of this PR with those in #6407, as mentioned in #6606?

Roxy-3D commented:
I'm worried about how intuitive this will be.  Especially for non-English people.   

SILENT doesn't convey what is being sought by the user with this option enabled.   Even in the fan case, the user is not trying to make the fan be quiet.  They are trying to get the fan to quit generating electrical noise.

bgort commented:
Yeah, that's a good point.  I guess unwanted EM is 'noise' to me, and you 'silence' noise to get rid of it, but that's not going to be understood by a good number of people...

At least renaming is easy enough.

Roxy-3D commented:
> At least renaming is easy enough.

Yeah...  Except when the change is in 37 Configuration.h files!   

How about we keep the first and add something like the second:
```cpp
#define BLTOUCH_HEATERS_OFF // Enable if the probe seems unreliable. Heaters will be disabled for each probe.
#define BLTOUCH_FANS_OFF  // if enabled fans will be shut down for each probe to reduce electrical noise
```

bgort commented:
I'm not opposed to that, but it's not just for BLTouch anymore -- it'll work with any probe type.

So maybe just
```
PROBING_HEATERS_OFF
PROBING_FANS_OFF
```
?

Roxy-3D commented:
Sorry....    Yes...   Your suggestion is better!!!!

bgort commented:
Now to see if I can manage a global rename/replace ...

Bob-the-Kuhn commented:
When I have to make similar changes I uses Astrogrep to list all the files, select all the files and then open them in Notepad++.  So much better than opening them individually.

bgort commented:
Thank you.

Pretty sure I can manage a 'replace in files' in the IDE, and that will grab everything as long as I first 'unexclude' the example_configurations directory from the index.  Will give it a shot in a few minutes.

Roxy-3D commented:
> When I have to make similar changes I uses Astrogrep to list all the files, select all the files and then open them in Notepad++. So much better than opening them individually.

That is a good idea!!!   I use Astrogrep all the time to hunt down things.   But I don't use its ability to kick off an editor.   That would make it much less painful to edit all the Configuration.h files.

Bob-the-Kuhn commented:
In Astrogrep I use the following setups to open the file(s) at the correct line:


![image](https://cloud.githubusercontent.com/assets/20692583/25782931/d63e818a-3318-11e7-92fb-3e9390d22e73.png)


![image](https://cloud.githubusercontent.com/assets/20692583/25782929/caf76b5c-3318-11e7-9356-720f026ebcc8.png)

Roxy-3D commented:
Next time I have to make changes in all the Configuration.h files...   I'll do this.  Except...  I'll tell it to start a Vim session.

bgort commented:
I was able to do it with a pseudo-regular expression, luckily:
![screenshot 2017-05-07 12 58 17](https://cloud.githubusercontent.com/assets/8062568/25783174/a789de18-3325-11e7-9781-6cc8cc9d037d.png)

Then I went back and did it again because I realized I forgot the // comments after each define, then updated my working configuration file manually so as to leave the old names behind so that I could then use the 'refactor' feature to rename the defines everywhere in the code, then did one last 'replace in path' with just the old defines to get the uses in comments.

I'm a -huge- fan of IDEs with global find & replace.

(BTW, this is a 30-days-free demo of CLion, which I think we can get a free annual license of for being an open source project...)

bgort commented:
Making some modifications to this to tighten up the time the heaters and fans are off.  If anyone is planning on testing, hold off a few...

bgort commented:
Simplified, reduced the amount of time things are paused by pausing at a lower level, tested thoroughly, etc.

Unless anyone has any last comments I'm comfortable merging this later tonight.  I think it addresses all of the issues we had, is safe, etc.

Roxy-3D commented:
Great!   I'll grab a fresh copy as soon as it is merged...

tcm0116 commented:
I'd like to see if we can't look into merging this with what I had put together in #6407 since I had added some capability for timing out the heaters and pausing them after a fixed amount of time, which is used by the M600 filament change. Another interesting thing I did was to flash the target temperatures in the LCD when the headers were paused as opposed to just showing 0. This had the nice benefit of indicating what the temperature will be after it is un-paused.

bgort commented:
I did see some areas where we could merge things.  

I really like the idea of flashing the temp when paused, but had missed that in my previous review of your PR.  I'll check it out again and then maybe we can put our heads together...
